File: typescript/src/executor/executor.ts
import type { Backend, Message } from "../llm/index.js";
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type { Response, ToolCall } from "../llm/response.js";
import { Role } from "../llm/index.js";
import type { CallerContext } from "../middleware/index.js";
import type { ToolRegistry } from "../tools/index.js";
import type { ToolFormatter } from "../toolformat/index.js";

export enum TerminationReason {
  COMPLETE = "complete",
  MAX_ITERATIONS = "max_iterations",
  ERROR = "error",
  CANCELED = "canceled",
}

export interface ExecuteRequest {
  system_prompt: string;
  user_message: string;
  history: Message[];
  max_iterations?: number;
  caller_ctx: CallerContext;
  job_id?: string;
}

export interface ExecuteResult {
  final_response: string;
  iterations: number;
  tool_calls_made: number;
  termination_reason: TerminationReason;
  conversation: Message[];
}

export interface Executor {
  execute(req: ExecuteRequest): ExecuteResult;
}

export interface InferenceExecutorConfig {
  backend: Backend;
  registry: ToolRegistry;
  tool_executor: unknown;
  formatter: ToolFormatter;
  max_iterations: number;
  completion_signal: string;
}

export class InferenceExecutor implements Executor {
  private config: InferenceExecutorConfig;

  constructor(config: InferenceExecutorConfig) {
    this.config = {
      ...config,
      max_iterations: config.max_iterations || 20,
      completion_signal: config.completion_signal || "TASK_COMPLETE",
    };
  }

  execute(req: ExecuteRequest): ExecuteResult {
    const conversation: Message[] = [...req.history];
    conversation.push({ role: Role.SYSTEM, content: req.system_prompt });
    conversation.push({ role: Role.USER, content: req.user_message });

    let iterations = 0;
    let tool_calls_made = 0;
    let final_response = "";

    const max_iters = req.max_iterations || this.config.max_iterations;

    while (iterations < max_iters) {
      const response = this.config.backend.complete(conversation);

      if (!response.tool_calls || response.tool_calls.length === 0) {
        final_response = response.content;
        break;
      }

      tool_calls_made += response.tool_calls.length;

      for (const tool_call of response.tool_calls) {
        const [result, success, error] = (this.config.tool_executor as {
          execute: (
            toolName: string,
            args: Record<string, unknown>,
            caller: CallerContext
          ) => [unknown, boolean, string];
        }).execute(tool_call.name, tool_call.arguments, req.caller_ctx);

        conversation.push({
          role: Role.TOOL,
          content: success
            ? `Tool ${tool_call.name} result: ${JSON.stringify(result)}`
            : `Tool ${tool_call.name} error: ${error}`,
          tool_call_id: tool_call.id,
        });
      }

      iterations++;
    }

    let termination: TerminationReason;
    if (iterations >= max_iters) {
      termination = TerminationReason.MAX_ITERATIONS;
    } else if (final_response) {
      termination = TerminationReason.COMPLETE;
    } else {
      termination = TerminationReason.ERROR;
    }

    return {
      final_response,
      iterations,
      tool_calls_made,
      termination_reason: termination,
      conversation,
    };
  }
}

File: python/src/pedro_agentware/executor/executor.py
"""Executor - Agent inference loop."""

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Protocol

from ..llm import Backend, Message
from ..llm.request import Role
from ..middleware import CallerContext
from ..toolformat import ToolFormatter
from ..tools import ToolRegistry


class TerminationReason(str, Enum):
    """Reason for executor termination."""

    COMPLETE = "complete"
    MAX_ITERATIONS = "max_iterations"
    ERROR = "error"
    CANCELED = "canceled"


@dataclass
class ExecuteRequest:
    """Input to a single agent run."""

    system_prompt: str
    user_message: str
    history: list[Message] = field(default_factory=list)
    max_iterations: int = 0
    caller_ctx: CallerContext = field(default_factory=CallerContext)
    job_id: str = ""


@dataclass
class ExecuteResult:
    """Output of a completed agent run."""

    final_response: str
    iterations: int
    tool_calls_made: int
    termination_reason: TerminationReason
    conversation: list[Message]


class Executor(Protocol):
    """Protocol for executors."""

    def execute(self, req: ExecuteRequest) -> ExecuteResult:
        """Run the inference loop for a task."""
        ...


@dataclass
class InferenceExecutorConfig:
    """Configuration for InferenceExecutor."""

    backend: Backend
    registry: ToolRegistry
    tool_executor: Any
    formatter: ToolFormatter
    max_iterations: int = 20
    completion_signal: str = "TASK_COMPLETE"


class InferenceExecutor:
    """Standard executor implementation."""

    def __init__(self, config: InferenceExecutorConfig) -> None:
        self._config = config

    def execute(self, req: ExecuteRequest) -> ExecuteResult:
        """Run the inference loop."""
        conversation = list(req.history)
        conversation.append(Message(role=Role.SYSTEM, content=req.system_prompt))
        conversation.append(Message(role=Role.USER, content=req.user_message))

        iterations = 0
        tool_calls_made = 0
        final_response = ""

        max_iters = req.max_iterations or self._config.max_iterations

        while iterations < max_iters:
            response = self._config.backend.complete(conversation)

            if not response.tool_calls:
                final_response = response.content
                break

            tool_calls_made += len(response.tool_calls)

            for tool_call in response.tool_calls:
                tool_name = tool_call.name
                tool_args = tool_call.arguments

                result, success, error = self._config.tool_executor.execute(
                    tool_name, tool_args, req.caller_ctx
                )

                conversation.append(
                    Message(
                        role=Role.TOOL,
                        content=f"Tool {tool_name} result: {result if success else error}",
                        tool_call_id=tool_call.id,
                    )
                )

            iterations += 1

        if iterations >= max_iters:
            termination = TerminationReason.MAX_ITERATIONS
        elif final_response:
            termination = TerminationReason.COMPLETE
        else:
            termination = TerminationReason.ERROR

        return ExecuteResult(
            final_response=final_response,
            iterations=iterations,
            tool_calls_made=tool_calls_made,
            termination_reason=termination,
            conversation=conversation,
        )
